_posts/2025-05-16-last-ten-years.markdown

```diff
@@ -0,0 +1,137 @@
+The [first public version of the CHERI ISA](https://www.cl.cam.ac.uk/techreports/UCAM-CL-TR-850.pdf) contained two instructions for sealing, `CSealCode` and `CSealData` (pages 97 and 99).
+These were intended to create cross-domain entry points.
+A cross-compartment call would be represented by two capabilities sealed with the same type.
+These would be installed in the PCC and an invoked-data capability register during a compartment call.
+
+By the 2015 paper, these were replaced by a single `CSeal` instruction that *could* seal pairs of capabilities in this way but could also seal capabilities independently.
+This let us implement opaque data types using sealed capabilities, something that we do extensively in CHERIoT (see [this section of the CHERIoT book for a longer explanation of CHERI object types and sealing](https://cheriot.org/book/concepts.html#sealing_intro)).
+I think our first public use of this (it was used a lot as an aside elsewhere) was our [2017 paper that extended type safety from Java code across the JNI boundary](https://www.cl.cam.ac.uk/research/security/ctsrd/pdfs/201704-asplos-cherijni.pdf).
+
+The `CCall` instruction at the time was actually a trap (with the expectation that a future version would move some or all of the trap handler into hardware).
+It branched to a handler, as did `CReturn`.
+The software stack in the paper used these to implement a trusted stack for cross-compartment calls.
+In CHERIoT, we use a very similar model but don't need a special instruction.
+We use the sealed entry (sentry) model, where one object type is reserved for capabilities that you can branch to but can't do anything else with.
+This was added a few years after the 2015 paper and meant that CHERIoT started from a model where the trap was not necessary.
+Using sentries means that our cross-compartment call can just use a jump to a sentry to invoke the switcher, rather than a trap.
+This is more scalable: a single system can provide multiple switchers like ours, but the 2015 paper required one per address space.
+In CHERIoT we extend sentries to control interrupt state and to differentiate between forward and backwards control-flow edges.
+
+If you look at the way that the ccall switcher works the 2015 paper, you'll see something that is quite recognisable from the CHERIoT compartment-switch routine.
+Both maintain a trusted stack and use it in similar ways.
+The paper expects each compartment to have its own stack (which made multithreading hard. Dapeng Gao [is doing exciting work here](https://www.cl.cam.ac.uk/research/security/ctsrd/pdfs/202306-plarch-library-based-compartmentalisation.pdf)), whereas CHERIoT reuses and zeroes the stack.
+The CHERIoT approach is optimised for small systems.
+Modern microcontrollers are fast but have small amounts of memory.
+Stacks are small and so zeroing them is a low overhead, whereas requiring a lot more stack would put more pressure on the most scarce resource in the system.
```

Comment:
Interesting to note that nothing about the ISA prohibits having multiple stacks if you want, although we did add some specialised hardware to help with stack zeroing.